openff/recharge/esp/qcarchive.py

```diff
@@ -0,0 +1,402 @@
+from typing import TYPE_CHECKING, Iterable, List, Optional, Tuple
+
+import numpy
+
+from openff.recharge.esp import DFTGridSettings, ESPSettings
+from openff.recharge.esp.storage import MoleculeESPRecord
+from openff.recharge.grids import GridGenerator
+from openff.recharge.utilities import requires_package
+from openff.recharge.utilities.exceptions import RechargeException
+from openff.recharge.utilities.openeye import molecule_to_conformers
+
+if TYPE_CHECKING:
+    import qcelemental.models
+    import qcelemental.models.results
+    import qcportal.models
+
+
+class MissingQCMoleculesError(RechargeException):
+    """An exception raised when an expected set of molecules are not present
+    in a QC data set."""
+
+    def __init__(self, data_set_name: str, missing_smiles: Iterable[str]):
+
+        smiles_string = "\n".join(missing_smiles)
+
+        super(MissingQCMoleculesError, self).__init__(
+            f"The {smiles_string} SMILES patterns were not found in the "
+            f"{data_set_name} data set."
+        )
+
+        self.data_set_name = data_set_name
+        self.missing_smiles = missing_smiles
+
+
+class MissingQCResultsError(RechargeException):
+    """An exception raised when an expected set of results are not present
+    in a QC data set."""
+
+    def __init__(self, data_set_name: str, missing_ids: Iterable[str]):
+
+        id_string = "\n".join(missing_ids)
+
+        super(MissingQCResultsError, self).__init__(
+            f"The result records associated with the following molecule ids from the "
+            f"{data_set_name} data set could not be retrieved from QCA: {id_string}"
+        )
+
+        self.data_set_name = data_set_name
+        self.missing_ids = missing_ids
+
+
+class MissingQCWaveFunctionError(RechargeException):
+    """An exception raised when a result does not store the required information about
+    a computed QM wavefunction."""
+
+    def __init__(self, result_id: str):
+
+        super(MissingQCWaveFunctionError, self).__init__(
+            f"The result with id={result_id} does not store the required wavefunction."
+            f"Make sure to use at minimum the 'orbitals_and_eigenvalues' wavefunction "
+            f"protocol when computing the data set."
+        )
+        self.result_id = result_id
+
+
+def _retrieve_qca_results(
+    data_set_name: str,
+    subset: Optional[Iterable[str]],
+    esp_settings: ESPSettings,
+    qcfractal_address: Optional[str] = None,
+) -> List[Tuple["qcelemental.models.Molecule", "qcportal.models.ResultRecord"]]:
+    """Attempt to retrieve the results for the requested data set from a QCFractal
+    server.
+
+    Parameters
+    ----------
+    data_set_name
+        The name of the data set to retrive the results from.
+    subset
```

Comment:
I like this having the ability to search for molecules explicitly rather than subset being the known indices you want to extract which is what this normally represents inqcportal.